### ISSUE TITLE
Add caching

### ISSUE THREAD

**Issue by wbraynen** [wbraynen/lighthouse-api#3]
Kinda [like this](https://codeforgeek.com/2015/07/using-redis-to-handle-session-in-node-js/), but maybe not on a per-session basis.  Maybe just pre-cache all the endpoints' json results and the serve them from the cache instead of from disk.

**Comment by wbraynen** [comment]
Will not merge into master.  The gains do not seem to be sufficiently significant in this case.

I used string caching, using `set` and `get` instead of `hset` and `hget`, which meant that to store a json I had to call `JSON.stringify` first to escape double quotes.  This in turn meant using `JSON.parse` to unmarshall the data.  (Not sure it would've been any faster using `hset` and `hget`; didn't try.)

With `JSON.parse`, this approach ended up being _slower_ (!) than not using redis at all.  (Does mongodb have a cache for reads?  It must, right?)

Without `JSON.parse`, it was indeed faster, but (a) I then get stuck with escaped double quotes in my jsons, so it's not very useful in that format and (b) it was still much less of a speedup than I expected, cutting average response time by only 12% from 257ms to 226ms:

![image](https://cloud.githubusercontent.com/assets/4765449/11091037/3dc7828e-8834-11e5-911b-552cb43d2ea1.png)

This performance test was conducted on localhost, on my laptop.  The results using `npm test`/mocha were similar.

**Comment by rexposadas** [comment]
MongoDB has has a cache layer. from the [docs](https://docs.mongodb.org/manual/faq/fundamentals/)

`MongoDB automatically uses all free memory on the machine as its cache.`

I took a glance at your `redis` branch. I would make sure that the redis warm-up is not called multiple times: https://github.com/wbraynen/lighthouse-api/blob/redis/app.js#L15

Also, I sometimes use the [ab](http://linux.die.net/man/1/ab) as a benchmarking tool. Using a benchmarking tool will help ensure that you have a consistent load on each test iteration.

**Comment by wbraynen** [comment]
Yeah, I saw that too.  I just don't know what it means.  As in, what kind of stuff does it cache?  As in, would redis simply be superfluous in my use case?

Re. redis warmup: `precacheCommonRoutes()` is called only once.  I can verify that by looking at the output from console.log().

I was timing it using `npm test` (which calls mocha, which outputs response times, which hits all the endpoints **once**), as well as by simply calling `time curl localhost:3000/nrc/datanotusedinrankings > /dev/null` **ten times** on the same endpoint (i.e. `/nrc/datanotusedinrankings`) and taking the mean, median and standard deviation.  It's the latter, the ten runs from calling `time` on `curl` on `/nrc/datanotusedinrankings` that you see in the table above, along with the three basic summary stats.

And cool!  I've never used Apache Benchmark to benchmark a rest api.  Off to check it out, right now : )

**Comment by wbraynen** [comment]
Ah cool!  ab lets you run concurrent tests.  So, I benchmarked it using `ab` (without concurrent runs, so it's the same kind of test as above) and this time taking a sample of 100 runs instead of 10, and the results are basically the same:

Firstly, the improvement isn't as much as I expected: only 14% from, on average, 232ms to 198ms.

But more importantly:

Redis with JSON.parse is slowest.
Redis without JSON.parse is fastest.  (This is the redis branch.)
Without redis, so just with mongo, is in the middle.  (This is the master branch.)

The problem is that I need to call JSON.parse because otherwise my API returns a json with escaped double quotes.  And that means that with redis my API's performance actually suffers if escaped double quotes are a deal breaker (which, I think they are).  So, I wonder if there is a better way to use redis or if, for some reason, redis just isn't suited to my use case (and if so, what's that reason?!).

### Details and reproducibility

The ab command I ran:

```
$ ab -n 100 localhost:3000/nrc/datanotusedinrankings
```

The results:

```
-------------------------
REDIS WITH JSON.PARSE: (redis branch with json.parse edit)

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.0      0       0
Processing:   243  278  25.5    273     358
Waiting:      240  275  25.3    270     356
Total:        243  278  25.5    273     359

-------------------------
WITHOUT REDIS.  MONGODB ONLY: (master branch)

              min  mean[+/-sd] median   max
Connect:        0    0   0.0      0       0
Processing:   200  232  24.3    229     338
Waiting:      197  229  24.2    227     335
Total:        200  232  24.3    229     338

-------------------------
REDIS WITHOUT JSON.PARSE: (redis branch)

              min  mean[+/-sd] median   max
Connect:        0    0   0.0      0       0
Processing:   147  197  36.0    188     350
Waiting:      144  195  35.9    186     347
Total:        147  198  36.0    189     350
```

**Comment by wbraynen** [comment]
Trying to use redis for jsons was a hack of course.  In short, for now looks like the only way to get a performance speedup with redis is to settle for this output:
![image](https://cloud.githubusercontent.com/assets/4765449/11155677/3a2a5170-8a04-11e5-88ec-f11a55930ac0.png)

Not sure what the right tool is to get further improvements though.  Maybe CouchDB or something else, or maybe there is a way to finetune/configure mongo's cache.  @rexposadas, any thoughts?
